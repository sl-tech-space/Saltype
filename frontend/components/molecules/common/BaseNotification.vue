<script setup lang="ts">
import Text from '~/components/atoms/texts/Text.vue';

interface Props {
    message: string;
    content: string;
    show: boolean;
}

const props = defineProps<Props>();
const isVisible = ref(props.show);
const notificationRef = ref<HTMLDivElement | null>(null);
let timer: ReturnType<typeof setTimeout> | null = null;

const showNotification = async () => {
    isVisible.value = true;
    await nextTick();
    adjustHeight();
    if (timer) clearTimeout(timer);
    timer = setTimeout(() => {
        isVisible.value = false;
    }, 5000);
};

const adjustHeight = () => {
    if (notificationRef.value) {
        notificationRef.value.style.height = 'auto';
        const height = notificationRef.value.scrollHeight;
        notificationRef.value.style.height = `${height}px`;
    }
};

watch(() => props.show, (newValue) => {
    if (newValue) {
        showNotification();
    } else {
        isVisible.value = false;
    }
}, { immediate: true });

onMounted(() => {
    if (props.show) {
        showNotification();
    }
});

onBeforeUnmount(() => {
    if (timer) clearTimeout(timer);
});
</script>

<template>
    <Transition name="fade">
        <div v-if="isVisible" ref="notificationRef" class="notification">
            <Text size="large" color="white">
                {{ message }}
            </Text>
            <Text color="white" v-html="content"></Text>
        </div>
    </Transition>
</template>


<style lang="scss" scoped>
.notification {
    width: 20%;
    min-height: 10vh;
    position: fixed;
    bottom: 20px;
    right: 20px;
    background-color: $black;
    padding: 15px;
    border: 2px solid $main-color;
    border-radius: 5px;
    box-shadow: 0 2px 10px $main-color;
    z-index: 1000;
    word-wrap: break-word;
    overflow: hidden;
    transition: height 0.3s ease;
}

.fade-enter-active,
.fade-leave-active {
    transition: opacity 0.5s, height 0.3s ease;
}

.fade-enter-from,
.fade-leave-to {
    opacity: 0;
}
</style>