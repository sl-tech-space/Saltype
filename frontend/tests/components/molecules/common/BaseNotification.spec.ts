import { describe, it, expect, vi, beforeEach } from "vitest";
import { mount, VueWrapper } from "@vue/test-utils";
import { nextTick } from "vue";
import BaseNotification from "../../../../components/molecules/common/BaseNotification.vue";

describe("BaseNotification", () => {
  let wrapper: VueWrapper<any>;

  beforeEach(() => {
    vi.useFakeTimers();
  });

  const mountNotification = (props = {}) => {
    return mount(BaseNotification, {
      props: {
        message: "Test Message",
        content: "Test Content",
        show: false,
        ...props,
      },
      global: {
        stubs: {
          Text: true,
          Transition: true,
        },
      },
    });
  };

  it("コンポーネントが正しくマウントされる", () => {
    wrapper = mountNotification();
    expect(wrapper.vm).toBeTruthy();
  });

  it("初期状態では通知が表示されない", () => {
    wrapper = mountNotification();
    expect(wrapper.find(".notification").exists()).toBe(false);
  });

  it("showプロパティがtrueになると通知が表示される", async () => {
    wrapper = mountNotification({ show: true });
    await nextTick();
    expect(wrapper.find(".notification").exists()).toBe(true);
  });

  /** 動作確認済み */
  // it("通知に正しいメッセージとコンテンツが表示される", async () => {
  //   wrapper = mountNotification({ show: true });
  //   await nextTick();
  //   const texts = wrapper.findAllComponents(Text);
  //   expect(texts[0].text()).toBe("Test Message");
  //   expect(texts[1].text()).toBe("Test Content");
  // });

  it("5秒後に通知が非表示になる", async () => {
    wrapper = mountNotification({ show: true });
    await nextTick();
    expect(wrapper.find(".notification").exists()).toBe(true);

    vi.advanceTimersByTime(5000);
    await nextTick();

    expect(wrapper.find(".notification").exists()).toBe(false);
  });

  it("高さが自動調整される", async () => {
    wrapper = mountNotification({ show: true });
    await nextTick();

    const notification = wrapper.find(".notification").element as HTMLElement;
    expect(notification.style.height).not.toBe("");
  });

  it("Transitionコンポーネントが正しく適用されている", () => {
    wrapper = mountNotification();
    expect(wrapper.findComponent({ name: "Transition" }).exists()).toBe(true);
  });

  /** CSR, SSRともに使用するためレンダリングの確認はしない */
  // it("コンポーネントが正しくレンダリングされる", async () => {
  //   wrapper = mountNotification();
  //   await nextTick();
  //   vi.runAllTimers();
  //   await nextTick();

  //   expect(wrapper.html()).toContain("notification");
  //   expect(wrapper.html()).toContain("Test Message");
  //   expect(wrapper.html()).toContain("Test Content");
  // });

  it("スナップショットと一致する", async () => {
    wrapper = mountNotification();
    await nextTick();
    vi.runAllTimers();
    await nextTick();

    expect(wrapper.html()).toMatchSnapshot();
  });
});
