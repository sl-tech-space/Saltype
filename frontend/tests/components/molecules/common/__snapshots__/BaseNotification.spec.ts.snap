// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BaseNotification > スナップショットと一致する 1`] = `
"<transition-stub data-v-977bef85="" name="fade" appear="false" persisted="false" css="true">
  <!--v-if-->
</transition-stub>"
`;
